src/QDL/usblib.js

```diff
@@ -0,0 +1,186 @@
+        }
+      }
+      this.maxSize = this.epIn.packetSize;
+    }
+    console.log("Endpoints: in =", this.epIn, ", out =", this.epOut);
+
+    try {
+        await this.device?.open();
+        await this.device?.selectConfiguration(1);
+        try {
+          await this.device?.claimInterface(0);
+        } catch(error) {
+          await this.device?.reset();
+          await this.device?.forget();
+          await this.device?.close();
+          throw error;
+        }
+    } catch (error) {
+      throw error;
+    }
+  }
+
+
+  async connect() {
+    console.log("Trying to connect Qualcomm device")
+    let devices = await navigator.usb.getDevices();
+    console.log("Found these USB devices:", devices);
+
+    this.device = await navigator.usb.requestDevice({
+      filters: [
+        {
+          vendorID  : vendorID,
+          productID : productID,
+          classCode : QDL_USB_CLASS,
+        },
+      ],
+    });
+    console.log("USing USB device:", this.device);
+
+    if (!this._registeredUsbListeners){
+
+      navigator.usb.addEventListener("disconnect", async (event) => {
+        if (event.device === this.device) {
+          console.log("USB device disconnected");
+          throw new Error("device was disconnected during flashing process");
+        }
+      });
+
+      navigator.usb.addEventListener("connect", async (event) =>{
+        console.log("USB device connect:", event.device);
+        this.device = event.device;
+
+        try {
+          await this._validateAndConnectDevice();
+        } catch (error) {
+          console.log("Error while connecting to the device");
+          throw error;
+        }
+      });
+
+      this._registeredUsbListeners = true;
+    }
+    await this._validateAndConnectDevice();
+  }
+
+
+  async read(resplen=null){
+    let respData = new Uint8Array();
+    let covered = 0;
+    if ((resplen === null)) 
+      resplen = this.epIn.packetSize;
+
+    while (covered < resplen) {
+      try {
+        let respPacket = await this.device?.transferIn(this.epIn?.endpointNumber, resplen);
+        respData = concatUint8Array([respData, new Uint8Array(respPacket.data.buffer)]);
+        resplen = respData.length;
+        covered += respData.length;
+      } catch (error) {
+        throw new Error("Errow while reading:", error)
+      }
+    }
+    return respData;
+  }
+
+  async write(cmdPacket, pktSize=null, wait=true, force=false) {
+    if (cmdPacket.length === 0 && force) {
+      try {
+        await this.device?.transferOut(this.epOut?.endpointNumber, cmdPacket);
+      } catch(error) {
+        try {
+          await this.device?.transferOut(this.epOut?.endpointNumber, cmdPacket);
+        } catch(error) {
+          throw new Error("Error while writing:", error);
+        }
+      }
+      return true;
+    }
+
+    let retry = 0;
+    let offset = 0;
+    if (pktSize === null)
+      pktSize = BULK_TRANSFER_SIZE;
+    while (offset < cmdPacket.length) {
```

Comment:
these while loops seem sketchy, suspect this is why it freezes sometimes

Reply:
We need this loop to send packets that has size larger than max usb bulk size. The reason it was freezing because I didn't catch error when disconnected (my bad on the implementation). Now it always throws error when disconnect (fail to read or write) --> never freeze when disconnect again.